### PR TITLE
Chores: using bulk delete if it's possible

### PR DESCRIPTION
**Pull request by dungdm93**
Using batch delete (a.k.a bulk delete) to perform delete multiple files if `FileIO` implement `SupportsBulkOperations`
#4012

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/util/FileIOUtil.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          LOG.warn("Failed to delete {} {}", e.numberFailedObjects(), name);
+        } catch (Exception e) {
+          // ignore
+        }
+      } else {
+        Tasks.foreach(files)
+            .noRetry()
+            .executeWith(service)
+            .suppressFailureWhenFinished()
+            .onFailure((file, exc) -> LOG.warn("Delete failed for {}: {}", name, file, exc))
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/util/FileIOUtil.java:88]: Are we missing a parameter?

File: core/src/main/java/org/apache/iceberg/util/FileIOUtil.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    public BulkDeleter executeWith(ExecutorService svc) {
+      this.service = svc;
+      return this;
+    }
+
+    public void execute() {
+      if (io instanceof SupportsBulkOperations) {
+        try {
+          SupportsBulkOperations bulkIO = (SupportsBulkOperations) io;
+          bulkIO.deleteFiles(files);
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/util/FileIOUtil.java:77]: Should we use the configured 'service' pool?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java [hunk 3]
```diff
@@ -678,11 +678,9 @@ public static List<SparkPartition> filterPartitions(
   }
 
   private static void deleteManifests(FileIO io, List<ManifestFile> manifests) {
-    Tasks.foreach(manifests)
+    FileIOUtil.bulkDeleteManifests(io, manifests)
```
Comment (szehon-ho) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:681]: Looks like we dont pass a name here.  Thinking about it, if a name is always recommended, why not make it mandatory parameter  of the bulkDelete API?

**Review by szehon-ho — COMMENTED** [review]
I think this is a good direction for me.  Also would like @amogh-jahagirdar to take a look to see if the FileIOUtil makes sense here or not.

**Comment by dramaticlly** [comment]
Also want to share a similar change for expire-snapshots to use BulkDeleton: https://github.com/apache/iceberg/pull/5412

File: core/src/main/java/org/apache/iceberg/util/FileIOUtil.java (hunk 1 above)
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/util/FileIOUtil.java:88]: Never mind about this.

**Comment by szehon-ho** [comment]
I think we need to coordinate with @amogh-jahagirdar who is also working on this on #5373.  There seems a few prs trying to use the bulk delete, and see if we can push the retry logic to the FileIO?

**Comment by dungdm93** [comment]
I don't think we should put retry logic into bulk delete. Because different client may wanna use different config. For examples `BaseTransaction` use `noRetry` while `HiveIcebergRecordWriter` use [`retry(3)`](https://github.com/apache/iceberg/blob/263441752393834c384a04d861cda1b8cb136a63/mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergRecordWriter.java#L101)

**Comment by amogh-jahagirdar** [comment]
@dungdm93 Not sure about comparing transactions and record writers but I think I understand what you're getting at. Another view on this (and I think this is what you are thinking, let me know if it's not) is that procedures should be in control of their retry strategy, failure handling and so on; not the FIleIO implementation because procedures/actions may want different configurations at different times.

My only doubt is that building a public Util which wraps around the existing Task framework mainly just for batch deletion seems heavy handed at this point in time. My feeling is that we shouldn't add public Util classes or any public kind of APIs unless we are really sure they will get used and so that's why I put it in the file IO implementation for simplicity; that seems more easily reversible in case folks want more configuration on retry behavior for different procedures. If we know that we want this customizability up front, then I think this makes sense. 

Would like to get the communities thoughts! @dungdm93 @szehon-ho @danielcweeks @rdblue

**Comment by szehon-ho** [comment]
> I don't think we should put retry logic into bulk delete. Because different client may wanna use different config. For examples BaseTransaction use noRetry while HiveIcebergRecordWriter use [retry(3)](https://github.com/apache/iceberg/blob/263441752393834c384a04d861cda1b8cb136a63/mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergRecordWriter.java#L101)

> @dungdm93 Not sure about comparing transactions and record writers but I think I understand what you're getting at. Another view on this (and I think this is what you are thinking, let me know if it's not) is that procedures should be in control of their retry strategy, failure handling and so on; not the FIleIO implementation because procedures/actions may want different configurations at different times.

To me, if we have decided to push down the batching logic S3FileIO.deleteFiles() which seems the case, I'm not sure I see another way than for the S3FileIO to handle the retry logic.  The caller cannot retry the whole thing in bulk, because some batches may succeeded and others failed, right?  If thats the case, the only thing we can do then is to have the callers pass in different options for retry to S3FileIO, to preserve their original intent?  For example expireSnapshot has retry (3), and these ones here have no retry, like you mentioned.  Or maybe I'm not being imaginative enough.

**Pull request closed without merging**